### ISSUE TITLE
FreeBSD install 

### ISSUE THREAD

**Issue by randomabuse** [andmarti1424/sc-im#93]
I was able to install under OS X 10.11 (El Capitan) , the only issue was installing the binary. I had to do that manually. Other than that, everything works great!

Same results with in Virtual Box with Fedora Linux - also very nice!

In contrast, my experiences with FreeBSD have been less than successful. 

When running 'make' from the 'src' directory, I am receiving what appears to be a linking error:

======================================================
cc   -o scim -lm 
/usr/lib/crt1.o: In function `_start':
/usr/src/lib/csu/amd64/crt1.c:(.text+0x17b): undefined reference to `main'
cc: error: linker command failed with exit code 1 (use -v to see invocation)
*** Error code 1

Stop.
make: stopped in /usr/home/tdurst/dev/git/sc-im/src

=======================================================

I have two FreeBSD servers (10.3) with console-only UI that I can test on and a PC-BSD/TrueOS desktop with a GUI.  

Any idea what I should be doing to get this working?

**Comment by andmarti1424** [comment]
Is that the complete error detail??
I dont see you're linking to ncurses library..
crt1.c is not a sc-im as well..
It's been a while since I don't build it in freebsd ..

**Comment by andmarti1424** [comment]
Please try as well removing optimization of gcc in Makefile..

**Comment by randomabuse** [comment]
Thank you for the reply and your suggestions. Very sorry about the delay, but will try to make up for it.

Yes, that was the complete error, it was pretty vague to me. You do seem to be on the right track in thinking it was a linking error.  

Removing the compiler optimization didn't help, though I also seem to be having some path issues; possibly unrelated. These only became apparent when I attempted to compile with clang (default compiler on FreeBSD 10.3).

I have some time to play with it today and will report my results.

Theodore

**Comment by styr0** [comment]
Any luck yet? I got similar errors on a PC-BSD box (FreeBSD 10.3-RELEASE), and more detail, but it's not clear what's missing.

It's definitely a linking issue. I compiled with gcc and clang, with/without optimizations, 32-/64-bit, no trouble for any of it (they both compile without errors or warnings).  I always get the same linker error, regardless of other flags like -nostartfiles. I've tried simpler ld commands. I've tried various library paths. It always defaults to  adding /usr/lib/crt*.o at this stage, which fail.

`/usr/lib/crt1.o: In function _start:`
`/tmp/build/H5zn/pcbsd/git/freebsd/lib/csu/amd64/crt1.c:(.text+0x16b):`
`undefined reference to main`

I'll have to look at it more closely and try building it on other platforms. I don't know enough about modern FreeBSD systems with the more complicated toolchains, but I would assume that a lot of FreeBSD ports must deal with issues like this now. I've not had time to sort it all out just yet : /

**Comment by andmarti1424** [comment]
Could not install freebsd yet.  Definitely is an issue that's related to the OS.. Did you try posting it in freebsd forum? From what I saw in Internet  other app builds showed the exact same error..

**Comment by andmarti1424** [comment]
Can you try building v0. 3 os sc-im? Instead of latest version?

**Comment by styr0** [comment]
@andmarti1424 Thanks!
I actually get the same error when I try v0.3.0 and v0.1.8 (I assume the same for other versions). This does look like something more OS-related, so I will check around some FreeBSD forums and porter documentation. If I find a good answer, or an easy fix for the makefile, I'll try posting it here.
Cheers

**Comment by andmarti1424** [comment]
That is weird since I used to build it in freebsd.. I will try to install freebsd with qemu and try it again. I do not use freebsd anymore. I believe the problem is is related as well..

**Comment by randomabuse** [comment]
Sorry for going dark on you guys, has been crazy busy at work.

I had the same results after toying with it for a while, so nothing new to report.

**Comment by andmarti1424** [comment]
Anyone reported this in FreeBSD forums?

**Comment by randomabuse** [comment]
Not I, sorry.

**Comment by andmarti1424** [comment]
I was able to get it done. these were my steps:


pkg install gmake
pkg install bison
pkg install git
pkg install gcc  (this installed binutils as well)
pkg install pkgconf
pkg install ncurses

# clone latest commit in depgraph branch..
git clone https://github.com/andmarti1424/sc-im

edit src/Makefile
and uncomment FREEBSD := -DFREEBSD line
and comment the LINUX one.

#build
gmake

#execute
./scim

The cursor was not visible though. This has to be because of the curses implementation.. Please report if you have this problem as well.

**Comment by randomabuse** [comment]
andmarti1424,

Confirmed those steps work on both 

* PC BSD/FreeBSD - Current (FreeBSD 12)
* FreeBSD 10.3

Things like 'gmake' need to be installed specifically and it does seem to make a difference here. I'm not a C developer, so am just guessing here.

The very welcome news is that this works now on FreeBSD.

**Comment by andmarti1424** [comment]
Did you build it with colors?

**Comment by randomabuse** [comment]
Color is working, but was a little washed out when I looked at it earlier today.

**Comment by styr0** [comment]
Hi,

Thanks for running that. I also got a crazy work schedule, so I didn't look at this again until today. It works now! My system already had all those packages, so I just had to use gmake instead of make.

There may be an even easier way to queue up the build in the FreeBSD ports tree if anyone wants to do the extra makefile(s) for that. I would now if I had the time... Anyway, this is plenty good enough.

Thanks again!
